enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go

```
@@ -182,3 +194,35 @@ func (s *sessionIssuerHelper) SessionData(token *oauth2.Token) oauth.SessionData
 		// TODO(beyang): store and use refresh token to auto-refresh sessions
 	}
 }
+
+// verifyUserGroups checks whether the authenticated user belongs to one of the GitLab groups when the allowGroups option is set
+func (s *sessionIssuerHelper) verifyUserGroups(ctx context.Context, glClient *gitlab.Client) bool {
+	var err error
+	var gitlabGroups []*gitlab.Group
+	hasNextPage := true
+
+	if len(s.allowGroups) == 0 {
+		return true
+	}
+
+	allowed := make(map[string]bool, len(s.allowGroups))
+	for _, group := range s.allowGroups {
+		allowed[group] = true
+	}
+
+	for page := 1; hasNextPage; page++ {
+		gitlabGroups, hasNextPage, err = glClient.ListGroups(ctx, page)
+		if err != nil {
+			return false
```

I think we need to print an error log before the `return`, otherwise the error is swallowed and going to be hard to debug. Or we should just make this method be able to return an error and propagate back to the client.

---

Good point!